memory CRUD silently degrades to a sql.js whole-image rename-over that can corrupt WAL databases under concurrent native writers

## Summary

Every memory CRUD operation in `@claude-flow/cli` (`storeEntry`, `getEntry`, `deleteEntry`, `purgeNamespace`, plus `ensureSchemaColumns` and `applyTemporalDecay`) is implemented bridge-first with a raw sql.js fallback: if the native better-sqlite3 bridge returns `null` for any reason, the operation silently reads the entire `.swarm/memory.db` image into an in-memory sql.js copy, mutates the copy, serializes the whole image with `db.export()`, and installs it via tmp-file + `rename()` over the live database path. That rename does not participate in SQLite's WAL/locking protocol at all. Under a concurrent native WAL writer (an `mcp start` server, a daemon, a hook-spawned CLI) it strands every already-open native connection on a deleted inode; committed-but-uncheckpointed transactions are omitted from the exported image and are no longer safely associated with the installed main image — they may be lost or misapplied — and the old `-wal`, which SQLite pairs to the main file by filename only, stays attached to a brand-new, unrelated image. sqlite.org documents rename/unlink-in-use and WAL mispairing as corruption mechanisms: this combination can cause acknowledged-write loss or corruption whenever native connections remain open. The demotion from native to sql.js is never logged — a registry import/initialize failure is cached process-wide, while a per-operation bridge exception silently falls back for that single call. `memory store` exits 0 even when it prints `[ERROR] database disk image is malformed`. The fallback fires on reads too: on the fallback path `memory retrieve` replaces the entire database image just to bump `access_count` (natively that bump is an ordinary incremental `UPDATE` — the defect is whole-image replacement on read, not that reads write). Entry into the fallback is deterministic with the package's own documented env switch, and once entered the path is provably unsafe: a minimal recipe reproduced acknowledged-write loss and index corruption in our runs, and we observed the identical signature on a live multi-session installation. The silent-demotion catch sites below are the reachable field entry conditions; we make no claim about measured field trigger frequency. A closely coupled destroy-path — `ensureInitialized`'s fresh-init clobber of a malformed-or-missing database — is included below. The proposed fix is capability-preserving: it keeps a working write path for genuinely-WASM-only hosts by gating the sql.js path on topology, rather than removing it.

## Environment

- `@claude-flow/cli` **3.32.8** as published on the npm registry. All `dist/src/...` line numbers below are from this published build's compiled JS.
- Node **v25.2.1**, Linux. better-sqlite3 11.10.0 in the npx cache loads and queries cleanly under this Node (compiled for `node_module_version: 141`, which matches `process.versions.modules`), so the fallback is not compensating for any ABI gap in this environment. The mechanism itself is filesystem-level, not Node-version-specific; for independent reproduction we suggest a current LTS Node (e.g. 22.x) with the pinned holder dependency below.
- Local **POSIX filesystem on real disk**. Filesystem semantics matter to the corruption mechanism: it depends on POSIX `rename(2)` atomically swapping the directory entry while open descriptors keep the old inode alive. Any POSIX filesystem is exposed. We have not tested Windows; see "Impact" for exactly what is and is not claimed there.
- Upstream `main` at `12ede2176` ships the same code: the only commits ahead of the 3.32.8-era base are the four #2721/#2725 plugin-hook commits, and `git log` over `src/memory`, `src/services`, and `src/fs-secure.ts` in that range is empty. `src/memory/memory-initializer.ts` at `12ede2176` still contains all seven `db.export()` whole-image sites (the `storeEntry` sql.js fallback begins at `memory-initializer.ts:2651`).

## Reproduction

Requires: `npx @claude-flow/cli@3.32.8`, `sqlite3` CLI, `npm i better-sqlite3@11.10.0` (pinned — the same engine and version the CLI's own bridge loads in this environment).

**Step 0 — confirm the routine engine is native WAL.** Verify via `PRAGMA journal_mode`; sidecar presence is corroborating but not guaranteed after the CLI exits (SQLite removes `-wal`/`-shm` when the last connection closes cleanly — in our runs they persisted, as shown; step 1 below re-checks them while a holder connection is open):

```bash
mkdir -p /tmp/cf-repro && cd /tmp/cf-repro
npx @claude-flow/cli@3.32.8 memory store --namespace repro --key seed --value seed
sqlite3 .swarm/memory.db 'PRAGMA journal_mode;'
# wal   <- routine engine is native SQLite in WAL mode
ls .swarm/
# memory.db  memory.db-wal  memory.db-shm   <- observed in our runs
```

Optionally populate more entries so the image is non-trivial (our run used a healthy ~16k-row database; any healthy DB reproduces).

**Step 1 — start a concurrent native WAL writer and keep it open.** Save as `holder.js` and run `node holder.js` in the repro directory:

```js
// holder.js — a concurrent native WAL writer, same engine the CLI's bridge uses
const Database = require('better-sqlite3');
const db = new Database('.swarm/memory.db');
db.pragma('journal_mode = WAL');
db.pragma('wal_autocheckpoint = 0'); // keep committed rows WAL-resident, as they are between checkpoints
db.exec('CREATE TABLE IF NOT EXISTS holder_rows (id INTEGER PRIMARY KEY, v TEXT)');
const ins = db.prepare('INSERT INTO holder_rows (v) VALUES (?)');
for (let i = 0; i < 50; i++) ins.run(`pre-${i}`); // 50 committed, WAL-only rows
console.log('50 rows committed; handle held open. Press Enter to write 20 more.');
process.stdin.once('data', () => {
  for (let i = 0; i < 20; i++) ins.run(`post-${i}`); // post-swap writes
  console.log('all inserts acknowledged');
  db.close();
  process.exit(0);
});
```

While the holder is running, `ls .swarm/` shows `memory.db-wal` and `memory.db-shm` present — a live native WAL attachment (this is the robust time to inspect sidecars, since they are removed on last clean close).

**Step 2 — trigger the sql.js fallback deterministically.** `CLAUDE_FLOW_DISABLE_BRIDGE=1` is the package's own switch (`dist/src/memory/memory-initializer.js:119-120`) and is a deterministic stand-in for *any* condition that makes the bridge return `null` in the wild:

```bash
stat -c 'inode=%i size=%s' .swarm/memory.db
CLAUDE_FLOW_DISABLE_BRIDGE=1 npx @claude-flow/cli@3.32.8 memory store --namespace repro --key trigger --value trigger
echo "exit=$?"                       # 0
stat -c 'inode=%i size=%s' .swarm/memory.db   # inode CHANGED
```

A `.memory.db.tmp-<pid>-<ts36>-<rand>` file appears in `.swarm/` and is renamed over the live `memory.db`. The store prints its `[OK]` acknowledgement. The holder process is still running with its handle open on the now-deleted old inode.

**Step 3 — post-swap native writes.** Press Enter in the holder terminal. In our run all 70/70 inserts (50 pre-swap + 20 post-swap) were acknowledged with zero errors — the native writer is completely oblivious that its main-DB descriptor points at a deleted file.

**Step 4 — fresh reader sees corruption.** After the holder exits, run **both** verification pragmas and attribute each recorded output to the exact command that produced it:

```bash
sqlite3 -readonly .swarm/memory.db 'PRAGMA quick_check;'
sqlite3 -readonly .swarm/memory.db 'PRAGMA integrity_check;'
```

In our runs the fresh reader reported `wrong # of entries in index idx_bridge_key` and `wrong # of entries in index idx_memory_key` — the package's own production indexes; in the field capture (root cause §5) `PRAGMA quick_check` itself emitted the same class of index-entry-count lines. (pragma.html documents `quick_check` as skipping the index↔table cross-check, so index-entry-count findings would ordinarily be expected from `integrity_check` — one more reason to capture both pragmas and label each output.) `integrity_check` is the comprehensive check (it additionally verifies UNIQUE constraints and index↔table consistency). Additionally, the row for `--key trigger` — acknowledged with `[OK]` in step 2 — is **gone** from the database.

**Step 5 — the failure self-perpetuates with success exit codes.** Run a plain store against the now-malformed DB (no env override needed — the bridge's native init now throws on the damaged image and the CLI silently falls back):

```bash
npx @claude-flow/cli@3.32.8 memory store --namespace repro --key after --value after
# prints: [ERROR] database disk image is malformed
echo $?   # 0 — and the inode swaps again: the corrupt bytes were read,
          # a row appended, and the whole corrupt image renamed back over the live path
```

**Control (shows concurrent native writers are otherwise handled correctly):** the same topology *without* the fallback is safe. With a healthy DB, one or two live native WAL writers, and five stock `memory store` invocations on the bridge path, we observed no inode change, clean integrity, and all rows durable. A store issued while another connection held a 20-second `BEGIN IMMEDIATE` lock simply waited out the lock and committed natively. Multi-process WAL concurrency works as designed when every writer is native — the sql.js whole-image fallback is the sole difference between "safe" and "destroyed."

## Observed vs Expected

**Observed**

- A memory CRUD op replaces the live database inode wholesale (`rename()` of a freshly exported image) while other processes hold open native WAL handles on it.
- Committed-but-uncheckpointed transactions in the `-wal` at the moment of the swap are omitted from the exported image and are no longer safely associated with the installed main image — they may be lost or misapplied by the stale-WAL overlay. Subsequent commits by already-open connections land on a deleted inode and are lost when those processes exit.
- A fresh reader that opens the new image finds the surviving old-generation `-wal` (paired by filename); in our runs it reported `wrong # of entries in index ...` / `database disk image is malformed`.
- The store's own `[OK]`-acknowledged row was lost in our runs.
- `memory retrieve` on the fallback path replaces the entire database image on a *read* — the `access_count` bump that is an ordinary incremental `UPDATE` on the native path.
- Failed stores print `[ERROR] database disk image is malformed` yet exit 0, and re-install the corrupt image, so the corruption round-trips forever.
- The native→sql.js demotion is never logged; a registry-level failure is cached for the life of the process, while a per-operation bridge error falls back silently for that single call (root cause §3).

**Expected**

- Mutations to a database that native connections hold open go through SQLite's WAL/locking protocol on the canonical inode, incrementally — never by exporting and renaming a whole image over a path other processes have open.
- A degraded (non-native) engine either operates only on a database it can prove is not native-attached — serialized against other degraded writers — or fails with a typed error and a nonzero exit code. It never silently substitutes an unsafe whole-image writer.
- Reads never replace the database image; a retrieval-time `access_count` bump is at most an incremental `UPDATE` through the engine's normal locking protocol.
- Any engine demotion is logged with its cause.

## Root cause analysis

All line numbers are from the published 3.32.8 `dist/src/...` JS.

**1. `memory store` is a per-invocation, daemon-unaware writer.** `commands/memory.js` `storeCommand.action` (`dist/src/commands/memory.js:83-163`) never contacts the workspace daemon or MCP; it directly imports and calls the writer (`:119-134`: `import('../memory/memory-initializer.js')` → `storeEntry(...)`). Every `npx @claude-flow/cli memory store` opens `.swarm/memory.db` itself, concurrently with whatever else holds it open.

**2. Bridge-first with a whole-image sql.js fallback.** `storeEntry` (`dist/src/memory/memory-initializer.js:2280-2387`) first tries the bridge (`getBridge()` at `:113-133` → `bridgeStoreEntry`, `memory-bridge.js:628+`), whose writes are normal incremental better-sqlite3 WAL inserts on the live inode (`memory-bridge.js:696-697`). If — and only if — the bridge returns `null`, the fallback (`memory-initializer.js:2299-2387`) runs:

```
readFileMaybeEncrypted(dbPath)        // read the WHOLE image, no locks, WAL-blind
new SQL.Database(fileBuffer)          // in-memory sql.js copy
db.run(insertSql, [...])              // mutate the copy
db.export()                           // serialize the WHOLE image      (line 2361)
writeFileRestricted(dbPath, ...)      // tmp + fsync + RENAME            (line 2362)
```

`writeFileRestricted` → `writeFileAtomic` (`fs-secure.js:106-121` → `:38-88`) writes `.${basename}.tmp-${pid}-${ts36}-${rand}` (`fs-secure.js:41`) then `renameSync(tmp, path)` (`fs-secure.js:50`). This exact tmp-name pattern is what we observed in the field (below).

The same shape exists in every steady-state mutator:

| Function | `db.export()` site (`dist/src/memory/memory-initializer.js`) | Trigger | Open-handle guard |
|---|---|---|---|
| `storeEntry` fallback | **2361** (write @ 2362) | every `memory store` when bridge unavailable | none |
| `getEntry` fallback | **2709** (write @ 2710) | every `memory retrieve` — the `access_count` bump (an ordinary incremental `UPDATE` on the native path) becomes a whole-image rewrite (`:2703-2710`) | none |
| `deleteEntry` fallback | **2838** | every `memory delete` | none |
| `purgeNamespace` fallback | **2978** | `memory purge` | `withMemoryDbLock` (`:2943`), an `O_EXCL <db>.lock` (`:2899-2931`) — honored only by other purges; no other writer or foreign process takes it |
| `ensureSchemaColumns` | **1027** | called at the top of the store/get fallbacks (`:2308`/`:2676`) | none |
| `applyTemporalDecay` | **1732** | TTL/decay maintenance | none |
| `initializeMemoryDatabase` | **1540** | fresh-DB creation — the one legitimate whole-image use | n/a |

Two native recovery paths (`recoverMemoryDatabase`, rename at `:1307` plus `-wal`/`-shm` deletion at `:1308-1313`; `restoreMemoryDbFromBackup`, rename at `dist/src/services/memory-backup.js:176`) also install whole files with no open-handle guard — they are addressed by the cooperative-drain design in Proposed fix item 3 rather than detailed here.

**3. The demotion is silent; one of its two paths is cached per-process.** Two distinct paths demote an operation to the whole-image fallback, and they cache differently:

- **Process-wide, cached demotion — registry import/initialize failure.** `bridgeStoreEntry` returns `null` whenever `getRegistry()` returns `null` (`memory-bridge.js:629-631`) — i.e. when `import('@claude-flow/memory')` fails or `registry.initialize(...)` throws. The catch at `memory-bridge.js:353-357` sets `bridgeAvailable = false` and the result is **cached for the life of the process**: every subsequent memory operation in that process takes the whole-image path.
- **Per-operation demotion — a `bridgeStoreEntry` exception.** `bridgeStoreEntry`'s own error handling is `catch { return null; }` (around `memory-bridge.js:741`): that single failing call returns `null` and that one operation falls back to sql.js. Nothing is cached; the next operation retries the bridge.

Nothing logs either transition. These catch sites are the reachable field entry conditions for the fallback; the reproduction's `CLAUDE_FLOW_DISABLE_BRIDGE=1` is a deterministic stand-in for them. We make no claim about how often they fire in the field — what is established is that the fallback is unsafe whenever it is entered while native connections are attached.

What flips the bridge on an *already-malformed* DB is clear, and it is the cached path: `registry.initialize` opens the damaged image, better-sqlite3 raises `database disk image is malformed`, the silent catch flips `bridgeAvailable=false`, and the CLI degrades to sql.js — which round-trips the damage (step 5 above). What triggered the *first* fallback on a previously healthy DB in our field incident is **unknowable post-hoc precisely because the demotion is silent** — that observability gap is part of the defect. We did rule out two candidates in this environment: it is not an ABI/load failure (better-sqlite3 loads and queries fine under Node v25.2.1) and not plain lock contention (bridge init succeeded even with an `IMMEDIATE` lock held by another handle; `busy_timeout` absorbs it). We suspect — labeled suspicion — that the first whole-image event on a healthy DB can also come from the init/migration/recovery whole-image writers in the table above, which run regardless of the store fallback.

**4. Why the rename can destroy a WAL database (SQLite primary sources).**

- **POSIX `rename(2)`** ([spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/rename.html)): "the removal of the file contents shall be postponed until all references to the file are closed." Already-open native connections keep their descriptors on the **old, now-unlinked inode** — they keep committing to a dead generation (our holder acknowledged 70/70 inserts), and new openers resolve the name to the new image.
- **[howtocorrupt.html §2.5](https://sqlite.org/howtocorrupt.html)** ("Unlinking or renaming a database file while in use") describes this exactly: "the two processes will be talking to different database files with the same name," and "Since rollback journals and WAL files are based on the name of the database file, the two different database files will share the same rollback journal or WAL file."
- **§1.2**: "it is important that any rollback journal (the \*-journal file) or write-ahead log (the \*-wal file) be copied together with the database file itself" — and copying is only safe "as long as there are no transactions in progress." The sql.js path reads the main image only, with no locks, mid-write.
- **§2.4 / §2.3**: "It is important that all connections to the same database file use the same locking protocol" — applications with different protocols "will not see each other's locks." sql.js is an in-memory WASM build with no OS file locking and no WAL/shared-memory VFS at all (our analysis of the failure class), so its read-and-replace is never serialized against native transactions.
- **§1.4**: corruption from "Overwriting a database file with another without also deleting any hot journal associated with the original database" — the WAL is the modern equivalent of the mispaired hot journal.
- **[fileformat2.html](https://sqlite.org/fileformat2.html)** — the load-bearing negative finding: a `-wal` is paired with its main database **by filename only**. A WAL frame is valid when its salts match the WAL's *own* header salts and the cumulative checksum chain holds; **no field in the main database header is validated against the WAL**. So a fresh opener of the sql.js-exported image finds the surviving old-generation `-wal`, its frames are self-consistent and therefore "valid," and SQLite overlays/checkpoints them onto the unrelated new image **page-by-page with no generation check** (our analysis of the documented format). [wal.html](https://sqlite.org/wal.html) corroborates the outcome class: "If a database file is separated from its WAL file, then transactions that were previously committed to the database might be lost, or the database file might become corrupted."
- A checkpoint against the mismatched pair then *cements* the corruption, since `wal_checkpoint` copies each valid frame to the main image by page number with no destination-generation verification (our analysis of [pragma.html](https://sqlite.org/pragma.html)).

**5. Field observation (live multi-session installation).** The `ruflo-cost-tracker` plugin's Stop hook — which fires at the end of **every** assistant turn in **every** session — spawns `npx @claude-flow/cli@latest memory store ...` (`scripts/track.mjs`, `spawnSync('npx', ['@claude-flow/cli@latest', 'memory', 'store', ...])`). We caught it in the act: a hook-spawned store wrote a fresh whole-database image to a `.memory.db.tmp-<pid>-<ts36>-<rand>` file (captured mid-write and still growing) and renamed it over the live `<workspace>/.swarm/memory.db` — the database inode changed between two stat captures taken shortly apart — while multiple long-lived `mcp start` processes each held an open descriptor on `memory.db (deleted)` plus descriptors on the still-current `-wal`/`-shm`. Multiple SQLite connections committing through different main-DB generations were sharing one live WAL/SHM. `PRAGMA quick_check` on that live DB (read-only), exactly as recorded (repeated index lines condensed in our record; the field capture is from `quick_check` only — the reproduction above records both pragmas):

```
Error: stepping, database disk image is malformed (11)
Tree 2 page 27816 cell 0: invalid page number 43298
Tree 2 page 1060 cell 7: overflow list length is 1 but should be 2
wrong # of entries in index idx_bridge_key / idx_bridge_ns
wrong # of entries in index idx_memory_owner/accessed/created/type/key/namespace
wrong # of entries in index sqlite_autoindex_memory_entries_1/2
wrong # of entries in index sqlite_autoindex_vector_indexes_1/2
```

Note the table b-tree damage (invalid page number, overflow-list error): this is not index-only corruption, so `REINDEX` cannot repair it; `.recover`-level salvage is the floor.

## Closely coupled destroy-path: `ensureInitialized` re-initializes over a malformed-or-missing database

Captured with a stack trace on 3.32.8: any MCP memory tool call routes through `ensureInitialized` (`dist/src/mcp-tools/memory-tools.js:211`, re-init call at `:217`), which treats a **missing or unreadable/malformed** database as uninitialized and calls `initializeMemoryDatabase` (`dist/src/memory/memory-initializer.js:1467`), which renames a fresh, schema-only (~0.5 MB) image over the existing path (`:1540-1542`: `db.export()` → `writeFileRestricted` → `writeFileAtomic` → `renameSync`). Consequences: (a) a malformed database — e.g. one produced by the fallback above — is silently **replaced by an empty one** on the next tool call, destroying content that `.recover` could otherwise salvage; (b) an MCP server that starts while the database file is briefly absent (backup/restore/maintenance windows) initializes a fresh tiny generation, and — reproduced on a fixture — its small-generation `-wal` then makes a full database restored at the same path read as `database disk image is malformed` to new readers, which in turn re-arms (a).

## Related observations (separate reports available)

- **PATH-drift in the plugin hook shim**: the shim that invokes the CLI on hook events resolves `ruflo`/`claude-flow` from PATH, so a stale global install can silently become the per-turn memory writer against the workspace database — version drift among writers multiplies any writer defect's blast radius.
- **`memory_stats` reads are WAL-blind on the sql.js path**: the stats read loads the main image only, so committed rows still resident in an un-checkpointed `-wal` are invisible and diagnostics under-count exactly when a native writer is active.

## Impact

- **Silent loss of acknowledged writes.** At the moment of the swap, committed-but-uncheckpointed transactions in the `-wal` are omitted from the exported image and no longer safely associated with the installed main image — they may be lost or misapplied; every subsequent commit by an already-open native connection lands on a deleted inode and evaporates when that process exits. Even the fallback store's own `[OK]`-acknowledged row was destroyed by the stale-WAL overlay in our reproduction.
- **Whole-database corruption**, including table b-tree damage that no `REINDEX` can repair — recovery from a corrupted image is `.recover`-level salvage plus backups.
- **Derived-state churn.** Each generation swap additionally strands the separately-persisted HNSW index — updated best-effort *after* the SQL acknowledgement with failures swallowed (`addToHNSWIndex`, `catch { return false }`; metadata written non-atomically via `fs.writeFileSync`, `memory-initializer.js:592-604`) — against a superseded database generation.
- **Self-perpetuating, with success exit codes.** Once the image is malformed, native bridge init throws on it, every subsequent CLI memory op silently takes the sql.js path, and each op re-installs the corrupt bytes while printing `[ERROR] database disk image is malformed` and exiting 0 — so hook and script callers cannot even detect the failure. In the field incident a freshly restored healthy database was re-corrupted by routine hook traffic alone.
- **Who is exposed (entry conditions, not a measured rate)**: any installation where more than one process touches `.swarm/memory.db` — which is the shipped default topology: long-lived `mcp start` servers hold native WAL handles while plugin hooks spawn transient `npx @claude-flow/cli@latest memory store` processes every turn, and concurrent sessions multiply the surface. One bridge demotion in any single process is sufficient to *enter* the unsafe path; we make no claim about how frequently that occurs in the field — what is demonstrated is that the path is unsafe whenever entered. Because `memory retrieve` also rewrites the image on the fallback path, "read-only" consumers are exposed too. The corruption mechanism we demonstrated is POSIX-rename-based (Linux/macOS). Upstream #2721/#2725 ("make ruflo-core/ruflo-cost-tracker hooks Windows-native", merged in `12ede2176`) mean the triggering Stop hook now *reaches* Windows; we have **not** demonstrated the rename mechanism there, and Windows file-sharing semantics typically prevent replacing a file that is open.

## Proposed fix

One design, capability-preserving: gate the degraded writer on topology instead of removing it, and make recovery cooperative. We have implemented and tested this shape as a reference implementation; every statement below is verified by tests, including real multi-process topologies.

1. **Native WAL mutation on the canonical inode as the primary path for every steady-state mutator.** `storeEntry` / `getEntry` / `deleteEntry` / `purgeNamespace` / `ensureSchemaColumns` / `applyTemporalDecay` open a better-sqlite3 handle on the live path (`journal_mode=WAL`, `busy_timeout`) and mutate incrementally — no image export, no rename. In our reference implementation the `db.export()` sites drop from 7 to 1 (only fresh-DB creation remains) and multi-process concurrent-writer tests pass with a stable inode. SQLite WAL is designed for this — our all-native control runs were fully safe under the same concurrency.
2. **A topology-gated sql.js path, so genuinely-WASM-only hosts keep write capability.** This is the *fallback tier* for hosts where native attachment is impossible; on mixed hosts the native path above is primary. When the native engine is unavailable, the sql.js read-modify-persist may still proceed — but only when the database is provably not native-attached: no `-wal`/`-shm` sidecars on disk and no live OS-level holders of the database, its WAL, or its SHM (in our reference implementation, a `/proc/<pid>/fd` scan on Linux; `lsof` on other POSIX platforms when present). While it proceeds, an advisory lock is held across the **entire** read-modify-persist cycle to serialize concurrent sql.js writers (a generalization of the existing `O_EXCL` purge lock at `:2899-2931`, which today only other purges honor). If sidecars or live holders are present — the exact topology that corrupts — the operation refuses with a typed, reasoned error and a nonzero exit. One necessary subtlety: the WAL-mode marker in the main-database header (the file-format write/read version bytes at offsets 18/19) is **not** by itself evidence of native attachment — sql.js sets those bytes itself when a stored schema runs `PRAGMA journal_mode=WAL`, and an at-rest, fully-checkpointed WAL-mode database with no sidecars and no holders is safe to image-mutate — so the gate must key on sidecars and holders, not header bytes.

   Honest limits of the gate, stated explicitly:
   - **TOCTOU residual**: a native opener can attach in the window between the assessment scan and the rename. No lock exists that both engines honor, so the gate is evidence-based, not race-free; on a genuinely WASM-only host no native opener exists and the race is moot, but on a mixed host it is a real (narrow) residual.
   - **Scan visibility**: the Linux `/proc` scan sees same-user processes in the current PID namespace (other users' fd tables are unreadable and skipped); the `lsof` path requires `lsof` to be installed. Holders outside that visibility are not detected.
   - **Unsupported platforms — proceed-with-warning, not fail-closed**: where neither scan mechanism is available, our reference implementation reports the scan unsupported and proceeds on the sidecar evidence alone, logging that it is proceeding unverified (the recovery paths print "cannot enumerate open handles on this platform — proceeding unverified"). This deliberately matches historical behavior on such platforms rather than failing closed; a stricter deployment can make refusal the policy there.
   - **Conservative refusals**: persisted `-wal`/`-shm` sidecars (e.g. left behind by a crashed native process) cause refusal even when no live holder exists — a deliberate false-refusal bias, since a leftover `-wal` may hold committed data the main image lacks.

   Our reference tests for this gate include a two-process race with no lost updates.
3. **Cooperative-drain recovery, so auto-recovery stays hands-off — among same-build accessors.** The whole-file installs that legitimately remain (`recoverMemoryDatabase`, `restoreMemoryDbFromBackup`, fresh-DB creation) run under a drain protocol instead of an unconditional rename: recovery posts a `<db>.recovery-pending` marker; cooperating connections detach themselves (new CLI openers wait before attaching; the long-lived bridge closes its handles and reattaches afterward); recovery bounded-waits for zero holders, performs the swap, and clears the marker. The drain can only ever apply to **same-build accessors that understand the marker protocol**; foreign or non-draining holders get a **typed refusal** — exactly the case where a swap would corrupt. Whole-file installs must also never delete `-wal`/`-shm` out from under live connections. Where a fresh consistent image must be produced from a readable database, prefer SQLite-supported mechanisms — the online backup API or `VACUUM INTO`, which read the logical current content through a live connection and are safe under concurrent writers — over raw file copies; the file-level swap remains only for images SQLite cannot read. Reference tests include unattended recovery completing under a real cooperating second OS process, and typed refusal against a real non-detaching holder.
4. **No silent demotion, no false exit codes, no writing reads.** Replace the bare `catch {}` / `catch { return null; }` swallowing in `memory-bridge.js` with a once-per-process log of the demotion and its cause — had this existed, the first trigger in our incident would be attributable; today it is structurally unknowable. `memory store` / `retrieve` / `delete` must exit nonzero when the operation failed or the database is malformed. `access_count` is bumped via a native `UPDATE` on the primary path, and on the sql.js path only inside the same gate-plus-lock — never as an unguarded whole-image rewrite. `ensureInitialized` must distinguish "no database" from "database exists but is unreadable/malformed" — the latter surfaces a typed error and never triggers re-initialization over the existing file (the coupled destroy-path above).

For completeness: the cleaner theoretical alternative — brokering all writes through a single native-attached process (e.g. the daemon) over IPC — is architecturally larger (the daemon today is a supervisor with no IPC write path), cannot help genuinely-WASM-only hosts, and is a compatible future enhancement that layers on top of this design rather than competing with it.

With items 1–3 in place, native hosts keep full write capability, WASM-only hosts keep write capability whenever the database is provably unattached, and self-healing remains unattended among same-build accessors. The guarantee is deliberately narrower than "no host class ever loses write capability": when safety cannot be established — native sidecars or holders present, or a holder that will not drain — the operation refuses with a typed error instead of proceeding, trading availability in exactly the topology where a whole-image rename can corrupt. Subject to the stated residuals (the assess-then-write race, scan visibility limits, and proceed-with-warning on scan-unsupported platforms), the corrupting topology is no longer reachable through any silent default path.

--- a/v3/@claude-flow/cli/__tests__/issue-2735-memory-wal-sidecar-guard.test.ts
+++ b/v3/@claude-flow/cli/__tests__/issue-2735-memory-wal-sidecar-guard.test.ts
@@ -1,0 +1,122 @@
+/**
+ * Regression coverage for issue #2735: memory CRUD's sql.js fallback did a
+ * whole-image read-modify-persist (export() + rename over the live
+ * database path) with no regard for whether a native better-sqlite3 WAL
+ * connection was already attached — corrupting the shared database or
+ * silently losing an acknowledged write when the fallback fired while a
+ * native writer (daemon, MCP server, another CLI invocation) held the file
+ * open.
+ *
+ * This suite covers the scoped fix actually shipped: refuse the sql.js
+ * fallback (`success: false`, typed error, never a whole-image write) when
+ * `-wal`/`-shm` sidecar files are present next to the database — strong
+ * evidence of a live native WAL connection, since a native connection keeps
+ * its sidecars on disk for its entire lifetime. It does NOT cover the
+ * fuller "scan live process holders" design also discussed in the issue —
+ * that remains a documented follow-up, not part of this fix.
+ *
+ * The bridge is force-disabled via CLAUDE_FLOW_DISABLE_BRIDGE=1 (the
+ * package's own documented switch, dist/src/memory/memory-initializer.js)
+ * so every case below exercises the sql.js fallback path directly,
+ * deterministically, without needing a real second native connection.
+ */
+import { afterEach, beforeEach, describe, expect, it } from 'vitest';
+import { mkdtempSync, rmSync, writeFileSync, existsSync } from 'node:fs';
+import { tmpdir } from 'node:os';
+import path from 'node:path';
+
+let dir: string;
+let dbPath: string;
+const ORIGINAL_ENV = process.env.CLAUDE_FLOW_DISABLE_BRIDGE;
+
+beforeEach(async () => {
+  dir = mkdtempSync(path.join(tmpdir(), 'ruflo-2735-'));
+  dbPath = path.join(dir, 'memory.db');
+  process.env.CLAUDE_FLOW_DISABLE_BRIDGE = '1';
+  // Reset the module registry so each test gets a fresh dynamic import —
+  // the module under test is pure-functional (no top-level state this
+  // suite depends on), but sql.js's own WASM init is safest re-run clean.
+  const { initializeMemoryDatabase } = await import('../src/memory/memory-initializer.js');
+  const initResult = await initializeMemoryDatabase({ dbPath, verbose: false });
+  expect(initResult.success).toBe(true);
+});
+
+afterEach(() => {
+  if (ORIGINAL_ENV === undefined) delete process.env.CLAUDE_FLOW_DISABLE_BRIDGE;
+  else process.env.CLAUDE_FLOW_DISABLE_BRIDGE = ORIGINAL_ENV;
+  rmSync(dir, { recursive: true, force: true });
+});
+
+describe('memory sql.js fallback WAL-sidecar guard — issue #2735', () => {
+  it('storeEntry refuses the whole-image write when -wal sidecar is present', async () => {
+    writeFileSync(`${dbPath}-wal`, '');
+    const { storeEntry } = await import('../src/memory/memory-initializer.js');
+    const result = await storeEntry({ key: 'k1', value: 'v1', dbPath });
+    expect(result.success).toBe(false);
+    expect(result.error).toMatch(/native WAL connection/i);
+  });
+
+  it('storeEntry refuses the whole-image write when -shm sidecar is present', async () => {
+    writeFileSync(`${dbPath}-shm`, '');
+    const { storeEntry } = await import('../src/memory/memory-initializer.js');
+    const result = await storeEntry({ key: 'k1', value: 'v1', dbPath });
+    expect(result.success).toBe(false);
+    expect(result.error).toMatch(/native WAL connection/i);
+  });
+
+  it('storeEntry proceeds normally with no sidecars present', async () => {
+    const { storeEntry } = await import('../src/memory/memory-initializer.js');
+    const result = await storeEntry({ key: 'k1', value: 'v1', dbPath, generateEmbeddingFlag: false });
+    expect(result.success).toBe(true);
+    expect(result.id).not.toBe('');
+  });
+
+  it('getEntry refuses the whole-image access_count-bump write when sidecars are present', async () => {
+    const { storeEntry } = await import('../src/memory/memory-initializer.js');
+    const stored = await storeEntry({ key: 'k2', value: 'v2', dbPath, generateEmbeddingFlag: false });
+    expect(stored.success).toBe(true);
+
+    writeFileSync(`${dbPath}-wal`, '');
+    const { getEntry } = await import('../src/memory/memory-initializer.js');
+    const result = await getEntry({ key: 'k2', dbPath });
+    expect(result.success).toBe(false);
+    expect(result.error).toMatch(/native WAL connection/i);
+  });
+
+  it('getEntry proceeds normally with no sidecars present', async () => {
+    const { storeEntry, getEntry } = await import('../src/memory/memory-initializer.js');
+    await storeEntry({ key: 'k3', value: 'v3', dbPath, generateEmbeddingFlag: false });
+    const result = await getEntry({ key: 'k3', dbPath });
+    expect(result.success).toBe(true);
+    expect(result.found).toBe(true);
+  });
+
+  it('deleteEntry refuses the whole-image write when sidecars are present', async () => {
+    const { storeEntry } = await import('../src/memory/memory-initializer.js');
+    await storeEntry({ key: 'k4', value: 'v4', dbPath, generateEmbeddingFlag: false });
+
+    writeFileSync(`${dbPath}-shm`, '');
+    const { deleteEntry } = await import('../src/memory/memory-initializer.js');
+    const result = await deleteEntry({ key: 'k4', dbPath });
+    expect(result.success).toBe(false);
+    expect(result.error).toMatch(/native WAL connection/i);
+  });
+
+  it('deleteEntry proceeds normally with no sidecars present', async () => {
+    const { storeEntry, deleteEntry } = await import('../src/memory/memory-initializer.js');
+    await storeEntry({ key: 'k5', value: 'v5', dbPath, generateEmbeddingFlag: false });
+    const result = await deleteEntry({ key: 'k5', dbPath });
+    expect(result.success).toBe(true);
+    expect(result.deleted).toBe(true);
+  });
+
+  it('a refused store never touches the database file (no whole-image write occurred)', async () => {
+    const before = existsSync(dbPath) ? require('node:fs').statSync(dbPath).mtimeMs : 0;
+    writeFileSync(`${dbPath}-wal`, '');
+    const { storeEntry } = await import('../src/memory/memory-initializer.js');
+    const result = await storeEntry({ key: 'k6', value: 'v6', dbPath });
+    expect(result.success).toBe(false);
+    const after = require('node:fs').statSync(dbPath).mtimeMs;
+    expect(after).toBe(before);
+  });
+});

--- a/v3/@claude-flow/cli/src/memory/memory-bridge.ts
+++ b/v3/@claude-flow/cli/src/memory/memory-bridge.ts
@@ -27,6 +27,29 @@ let registryPromise: Promise<any> | null = null;
 let registryInstance: any = null;
 let bridgeAvailable: boolean | null = null;
 
+// #2735 — the native-bridge -> sql.js whole-image fallback used to be
+// completely silent: a registry init failure cached `bridgeAvailable =
+// false` for the rest of the process, and a per-operation bridge exception
+// swallowed itself with a bare `catch { return null; }` — in both cases the
+// caller fell back to memory-initializer.ts's unsafe whole-image sql.js
+// path with zero diagnostic trail. When that path corrupted a database, the
+// demotion that caused it was structurally unknowable after the fact. Log
+// once per distinct reason (not once per call — a hot per-op catch could
+// otherwise spam stderr on every single memory operation) so the first
+// occurrence is attributable. Stderr only — never stdout, which callers may
+// be parsing as JSON.
+const loggedDemotions = new Set<string>();
+function logDemotionOnce(where: string, reason: string): void {
+  const dedupeKey = `${where}:${reason}`;
+  if (loggedDemotions.has(dedupeKey)) return;
+  loggedDemotions.add(dedupeKey);
+  try {
+    process.stderr.write(
+      `[memory-bridge] demoted to sql.js fallback in ${where}: ${reason}\n`,
+    );
+  } catch { /* stderr unavailable — nothing more we can do */ }
+}
+
 /**
  * Resolve database path with path traversal protection.
  * Only allows paths within or below the project's working directory,
@@ -325,9 +348,13 @@ async function getRegistry(dbPath?: string): Promise<any | null> {
         registryInstance = registry;
         bridgeAvailable = true;
         return registry;
-      } catch {
+      } catch (err) {
         bridgeAvailable = false;
         registryPromise = null;
+        logDemotionOnce(
+          'getRegistry (process-wide, cached for the rest of this process)',
+          err instanceof Error ? err.message : String(err),
+        );
         return null;
       }
     })();
@@ -795,7 +822,8 @@ export async function bridgeStoreEntry(options: {
       cached: true,
       attested: true,
     };
-  } catch {
+  } catch (err) {
+    logDemotionOnce('bridgeStoreEntry', err instanceof Error ? err.message : String(err));
     return null;
   }
 }
@@ -1158,7 +1186,8 @@ export async function bridgeGetEntry(options: {
     await cacheSet(registry, cacheKey, entry);
 
     return { success: true, found: true, cacheHit: false, entry };
-  } catch {
+  } catch (err) {
+    logDemotionOnce('bridgeGetEntry', err instanceof Error ? err.message : String(err));
     return null;
   }
 }
@@ -1234,7 +1263,8 @@ export async function bridgeDeleteEntry(options: {
       remainingEntries: remaining,
       guarded: true,
     };
-  } catch {
+  } catch (err) {
+    logDemotionOnce('bridgeDeleteEntry', err instanceof Error ? err.message : String(err));
     return null;
   }
 }
@@ -1298,7 +1328,8 @@ export async function bridgePurgeNamespace(options: {
       remainingEntries: remaining,
       guarded: true,
     };
-  } catch {
+  } catch (err) {
+    logDemotionOnce('bridgePurgeNamespace', err instanceof Error ? err.message : String(err));
     return null;
   }
 }

--- a/v3/@claude-flow/cli/src/memory/memory-initializer.ts
+++ b/v3/@claude-flow/cli/src/memory/memory-initializer.ts
@@ -37,6 +37,33 @@ function isRuvectorCoreResolvable(): boolean {
 }
 
 /**
+ * #2735 — before a whole-image sql.js read-modify-persist (export() +
+ * rename over the live database path), refuse if there is evidence of a
+ * live native (better-sqlite3) WAL connection: `-wal`/`-shm` sidecar files
+ * on disk. A native connection in WAL mode keeps its sidecars present for
+ * its entire lifetime (removed only on the last connection's clean close),
+ * so their presence is strong evidence of a live native holder — and their
+ * absence means the image is a clean, checkpointed, standalone file that
+ * sql.js can safely read-modify-write.
+ *
+ * This is a scoped-down version of the fuller "scan live process holders"
+ * design discussed in #2735: it does not close the narrow assess-then-write
+ * race (a native opener could still attach in the gap between this check
+ * and the write), but it directly closes the demonstrated corruption
+ * mechanism — a whole-image write proceeding while an ALREADY-OPEN native
+ * connection's sidecars are on disk — with no platform-specific process
+ * scanning. Fails closed (treats a stat error as "unsafe") because this is
+ * a safety gate, not a best-effort probe.
+ */
+function hasNativeWalSidecars(dbPath: string): boolean {
+  try {
+    return fs.existsSync(`${dbPath}-wal`) || fs.existsSync(`${dbPath}-shm`);
+  } catch {
+    return true;
+  }
+}
+
+/**
  * #1854: previously every site that needed the memory directory hardcoded
  * `getMemoryRoot()`, so the documented config entry
  * points (`memory.persistPath` config field, `memory configure --path`,
@@ -2668,6 +2695,23 @@ export async function storeEntry(options: {
       return { success: false, id: '', error: 'Database not initialized. Run: claude-flow memory init' };
     }
 
+    // #2735 — refuse an unsafe whole-image write while a native WAL
+    // connection appears to be attached (sidecars on disk). See
+    // hasNativeWalSidecars()'s doc comment for the corruption mechanism
+    // this closes and its known residual (the narrow assess-then-write
+    // race). This check gates ensureSchemaColumns()'s own whole-image
+    // write below too, not just this function's.
+    if (hasNativeWalSidecars(dbPath)) {
+      return {
+        success: false,
+        id: '',
+        error: 'memory database has an active native WAL connection ' +
+          '(found -wal/-shm sidecar files) — refusing an unsafe sql.js ' +
+          'whole-image write. Retry once the native writer completes, or ' +
+          'restore the native better-sqlite3 bridge.',
+      };
+    }
+
     // Ensure schema has all required columns (migration for older DBs)
     await ensureSchemaColumns(dbPath);
 
@@ -3179,6 +3223,21 @@ export async function getEntry(options: {
       return { success: false, found: false, error: 'Database not found' };
     }
 
+    // #2735 — see storeEntry's identical gate for the corruption mechanism
+    // this closes. Applies here too because the fallback's access_count
+    // bump is itself a whole-image write, not a lightweight read, even
+    // though this function's contract reads as a "get".
+    if (hasNativeWalSidecars(dbPath)) {
+      return {
+        success: false,
+        found: false,
+        error: 'memory database has an active native WAL connection ' +
+          '(found -wal/-shm sidecar files) — refusing an unsafe sql.js ' +
+          'whole-image read/write. Retry once the native writer completes, ' +
+          'or restore the native better-sqlite3 bridge.',
+      };
+    }
+
     // Ensure schema has all required columns (migration for older DBs)
     await ensureSchemaColumns(dbPath);
 
@@ -3317,6 +3376,22 @@ export async function deleteEntry(options: {
         namespace,
         remainingEntries: 0,
         error: 'Database not found'
+      };
+    }
+
+    // #2735 — see storeEntry's identical gate for the corruption mechanism
+    // this closes.
+    if (hasNativeWalSidecars(dbPath)) {
+      return {
+        success: false,
+        deleted: false,
+        key,
+        namespace,
+        remainingEntries: 0,
+        error: 'memory database has an active native WAL connection ' +
+          '(found -wal/-shm sidecar files) — refusing an unsafe sql.js ' +
+          'whole-image write. Retry once the native writer completes, or ' +
+          'restore the native better-sqlite3 bridge.',
       };
     }
 

